src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoOfTClassTProperty.cs

```diff
@@ -0,0 +1,194 @@
+    {
+        private bool _isPropertyPolicy;
+        internal Func<TClass, TProperty> Get { get; private set; }
+        internal Action<TClass, TProperty> Set { get; private set; }
+
+        // Constructor used for internal identifiers
+        internal JsonPropertyInfo() { }
+
+        internal JsonPropertyInfo(Type classType, Type propertyType, PropertyInfo propertyInfo, Type elementType, JsonSerializerOptions options) :
+            base(classType, propertyType, propertyInfo, elementType, options)
+        {
+            if (propertyInfo != null)
+            {
+                if (propertyInfo.GetMethod?.IsPublic == true)
+                {
+                    HasGetter = true;
+                    Get = (Func<TClass, TProperty>)Delegate.CreateDelegate(typeof(Func<TClass, TProperty>), propertyInfo.GetGetMethod());
+                }
+
+                if (propertyInfo.SetMethod?.IsPublic == true)
+                {
+                    HasSetter = true;
+                    Set = (Action<TClass, TProperty>)Delegate.CreateDelegate(typeof(Action<TClass, TProperty>), propertyInfo.GetSetMethod());
+                }
+            }
+            else
+            {
+                _isPropertyPolicy = true;
+                HasGetter = true;
+                HasSetter = true;
+            }
+
+            GetPolicies(options);
+        }
+
+        internal override object GetValueAsObject(object obj, JsonSerializerOptions options)
+        {
+            if (_isPropertyPolicy)
+            {
+                return obj;
+            }
+
+            Debug.Assert(Get != null);
+            return Get((TClass)obj);
+        }
+
+        internal override void SetValueAsObject(object obj, object value, JsonSerializerOptions options)
+        {
+            Debug.Assert(Set != null);
+            TProperty typedValue = (TProperty)value;
+
+            if (typedValue != null || !IgnoreNullPropertyValueOnWrite(options))
+            {
+                Set((TClass)obj, (TProperty)value);
+            }
+        }
+
+        internal override void Read(JsonTokenType tokenType, JsonSerializerOptions options, ref ReadStack state, ref Utf8JsonReader reader)
+        {
+            if (ElementClassInfo != null)
+            {
+                // Forward the setter to the value-based JsonPropertyInfo.
+                JsonPropertyInfo propertyInfo = ElementClassInfo.GetPolicyProperty();
+                propertyInfo.ReadEnumerable(tokenType, options, ref state, ref reader);
+            }
+            else if (HasSetter)
+            {
+                if (ValueConverter != null)
+                {
+                    Type propertyType = PropertyType;
+                    if (propertyType.IsGenericType && propertyType.GetGenericTypeDefinition() == typeof(Nullable<>))
+                    {
+                        propertyType = Nullable.GetUnderlyingType(propertyType);
```

Comment:
Perhaps cache the non-nullable property type so it's not computed on each property being read?